### PR TITLE
Packaging guide

### PR DESCRIPTION
**Pull request by kn0wmad**
WIP - initial thoughts requested.  I think I will be editing this quite heavily as I was shooting in the dark a bit.

### CODE REVIEW

File: ServicePackaging.md [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+echo start9/public > .backupignore
+echo start9/shared >> .backupignore
+
+home_type=$(yq e '.homepage.type' start9/config.yaml)
+subdomains=($(yq e '.subdomains.[].name' start9/config.yaml))
+```
+
+7. First we've defined the file as a bash, exported the IP address, chosen folders to ignore during backup, and set the homepage and subdomains types from our config.
+
+***I'VE ABANDONED THIS SECTION AFTER DISCUSSION LEADING TO THE IDEA THAT THIS MAY NOT BE A GOOD EXAMPLE FILE***
```
Comment (chrisguida) [inline ServicePackaging.md:248]: 🤣

Comment (chrisguida) [reply ServicePackaging.md:248]: Perhaps `cups` would be better?

File: ServicePackaging.md [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+	embassy-sdk pack
+
+image.tar: Dockerfile docker_entrypoint.sh
+	DOCKER_CLI_EXPERIMENTAL=enabled docker buildx build --tag start9/embassy-pages/main:0.1.3 --platform=linux/arm/v7 -o type=docker,dest=image.tar .
+```
+
+1. The first line simply removes the progress of a `make` process if it fails.
+`.DELETE_ON_ERROR:`
+
+2. The `all` step is run when the `make` command is issued.  This means that the `embassy-pages.s9pk` must first be built, and so on.
```
Comment (chrisguida) [inline ServicePackaging.md:393]: The Makefile format has changed substantially in 030, see [lnd-wrapper for 0.3.0](https://github.com/Start9Labs/lnd-wrapper/pull/24/files)

File: ServicePackaging.md [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+
+4. Next the `.s9pk` is created with the `embassy-sdk pack` command, supplied with the `manifest`, `config_spec`, previously created `image.tar`, and `instructions.md`.  Your project may likely also contain a `config_rules` file.  Some of these files we have not yet edited, but that will come shortly.
+
+For more details on creating a `Makefile` for your project, please check [here](https://docs.start9.com/contributing/services/makefile.html).
+
+### Service Config Specification and Rules
+
+Most self-hosted packages require a configuration.  With EmbassyOS, these config options are provided to the user in a friendly GUI, and invalid configs are not permitted.  This allows the user to manage their software without a lot of technical skill.  Two files are created in this process:
+
+`config_spec.yaml` for specifying all the config options your package depends on to run
```
Comment (chrisguida) [inline ServicePackaging.md:405]: This file and `config_rules.yaml` moved to the `assets/compat/` directory for 0.2.x compatibility, though in general 030's configs are much more flexible and are actually a getter and setter defined by the wrapper

File: ServicePackaging.md (hunk 1 above)
Comment (chrisguida) [reply ServicePackaging.md:248]: On second thought we should probably just update the `hello-world` wrapper with all of the features it's missing, like dependencies

**Review by chrisguida — COMMENTED** [review]
This is probably pretty good for now. The stuff that's missing is mostly stuff that's still being ironed out.

File: ServicePackaging.md [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+At minimum you will need the following:
+    1. [docker](https://docs.docker.com/get-docker)
+    2. [docker-buildx](https://docs.docker.com/buildx/working-with-buildx/)
+    3. ***PLACEHOLDER FOR EOS-SDK***
+
+The following are recommended:
+    4. [cargo](https://doc.rust-lang.org/cargo/)
+    5. [yq](https://mikefarah.gitbook.io/yq/) (version 4)
+    6. [make](https://www.gnu.org/software/make/)
+    7. [rust-musl-cross](***PLACEHOLDER FOR NEW MUSL-CROSS REPO***) (If using an Alpine image, which is highly recommended)
```
[the remark this answers is not in the document]
Comment (kn0wmad) [reply ServicePackaging.md:29]: Addressed

File: ServicePackaging.md [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+
+## Overview
+
+### Components
+
+Simply, the package is made up of the following parts:
+    1. Image - Each service is running in a Docker image.  Best results will come from an arm based linux; [Alpine](https://www.alpinelinux.org/) is highly recommended.
+    2. Volume - Each service gets a volume, allocated by EOS.  The volume directory (for seeding data into the volume) is typically: /root/volumes/<service-id>
```
[the remark this answers is not in the document]
Comment (kn0wmad) [reply ServicePackaging.md:37]: Fixed

File: ServicePackaging.md [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+    3. Dependencies - Rules and requirements of your service, which appear as simple UI elements, such as inputs, toggles, and drop-downs.  These are enforced by validations and clear user instructions.  EmbassyOS has a unique and powerful system for managing dependencies which allows anyone to have the power of systems administrators without the advanced skillset.
+    4. Manifest - Describes the service and its requirements.  This is for the marketplace listing, install considerations, license, donation address, and dependency requirements, and additional info.
+    5. Config - EOS makes a service's configuration available to the user in the GUI and must be valid regardless of user skill.
+    6. .s9pk Bundle - The config and manifest files get bundled into a .s9pk package.  This is the file a user downloads from the Marketplace, at which point EOS un-packages and installs the service.
```
[the remark this answers is not in the document]
Comment (kn0wmad) [reply ServicePackaging.md:41]: Fixed

File: ServicePackaging.md [hunk 7]
```diff
@@ -0,0 +1,533 @@
...
+git clone https://github.com/Start9Labs/hello-world-wrapper
+mv hello-world-wrapper embassy-pages-wrapper && cd embassy-pages-wrapper
+```
+
+2. Edit the `README.md` to explain what the service is, what dependencies are required, and build/install instructions.
+
+3. Edit the `manifest` file.  This must be in `.json`, `.toml`, or `.yaml` format and in `kebab-case` style.  You can see our Embassy Pages example `manifest.toml` below:
+
+```
+id = "embassy-pages"
```
[the remark this answers is not in the document]
Comment (kn0wmad) [reply ServicePackaging.md:71]: Linked to hello-world manifest, will need to update

File: ServicePackaging.md [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+image = "compat"
+system = true
+entrypoint = "true"
+args = []
+mounts = {}
+```
+
+Note the `[dependencies]` and `[volumes]` sections, which include File Browser such that files stored on a user's Embassy can be accessed and hosted onto a Page.
+
+For details on all the different possible dependency, type, and subtype definitions available for the `manifest` file, please see [here](https://docs.start9.com/contributing/services/manifest.html).
```
[the remark this answers is not in the document]
Comment (kn0wmad) [reply ServicePackaging.md:164]: Yeah there will be several updates to the existing service packaging guide.  I'll go through that after this doc is complete and get it up to date

File: ServicePackaging.md (hunk 1 above)
Comment (kn0wmad) [reply ServicePackaging.md:248]: Leaning toward hello-world updates being the best course of action

File: ServicePackaging.md (hunk 2 above)
Comment (kn0wmad) [reply ServicePackaging.md:393]: I think will move to hello-world, and my next project probably ought to be to get that updated for 0.3.0

File: ServicePackaging.md (hunk 3 above)
Comment (kn0wmad) [reply ServicePackaging.md:405]: I'll have to address this when I get hello-world updated and can rewrite this at that time

**Comment by kn0wmad** [comment]
WIP, to be rewritten to include `hello-world` after it has been updated for v0.3.0

**Pull request closed without merging**